rules/CodeQuality/Rector/Foreach_/SimplifyForeachToArrayFilterRector.php

```diff
@@ -119,6 +127,30 @@ private function shouldSkip(Foreach_ $foreach): bool
         return $ifNode->elseifs !== [];
     }
 
+    private function shouldSkipForeachKeyUsage(If_ $ifNode, Expr $keyVar): bool
+    {
+        if (! $keyVar instanceof Variable) {
+            return false;
+        }
+
+        $keyVarUsage = $this->betterNodeFinder->find(
+            $ifNode,
+            fn (Node $node): bool => $this->exprUsedInNodeAnalyzer->isUsed($node, $keyVar)
+        );
+
+        $keyVarUsageCount = count($keyVarUsage);
+        if ($keyVarUsageCount === 1) {
+            $parentArrayDimFetch = $keyVarUsage[0]->getAttribute(AttributeKey::PARENT_NODE);
+            if (! $parentArrayDimFetch instanceof ArrayDimFetch) {
+                return true;
+            }
+
+            return $parentArrayDimFetch->dim !== $keyVarUsage[0];
```

Comment:
I tried locally, you can actually use `ReadExprAnalyzer` service:

```suggestion
            return ! $this->readExprAnalyzer->isExprRead(current($keyVarUsage));
```

You need to inject the controller with `Rector\ReadWrite\NodeAnalyzer\ReadExprAnalyzer` service.

Reply:
@Wohlie I updated to use `ReadExprAnalyzer` service to your branch https://github.com/rectorphp/rector-src/pull/3100/commits/76fa68122fab6c3e97aa8b0e55534778d97e7555